contrib/babelfishpg_tsql/src/pl_exec.c

```diff
@@ -5458,71 +5445,67 @@ pltsql_update_identity_insert_sequence(PLtsql_expr *expr)
 			if (tuptable != NULL && n_processed > 0)
 			{
 				TupleDesc tupdesc_ret = tuptable->tupdesc;
-				HeapTuple tuple = tuptable->vals[n_processed-1];
 
-				/* Obtain the user identity value */
+				/* Obtain the user identity column */
 				for (attnum = 0; attnum < tupdesc_ret->natts; attnum++)
 				{
 					Form_pg_attribute attr = TupleDescAttr(tupdesc, attnum);
-					bool isnull;
 
 					/* Find by name since other attributes not defined */
 					if (strcmp(NameStr(attr->attname), id_attname) == 0)
 					{
-						int64 last_identity = DatumGetInt64(SPI_getbinval(tuple,
-																		  tupdesc_ret,
-																		  attnum+1,
-																		  &isnull));
+						int tup_idx;
+						int64 last_identity;
+						int64 min_identity = INT_MAX;
+						int64 max_identity = INT_MIN;
+						ListCell *seq_lc;
+						List *seq_options;
+
+						for (tup_idx = 0; tup_idx < n_processed; tup_idx++)
+						{
+							bool isnull;
+							HeapTuple tuple = tuptable->vals[tup_idx];
+
+							last_identity = DatumGetInt64(SPI_getbinval(tuple,
+																		tupdesc_ret,
+																		attnum+1,
+																		&isnull));
+
+							if (last_identity < min_identity)
+								min_identity = last_identity;
+							if (last_identity > max_identity)
+								max_identity = last_identity;
+						}
 
+						/* update last used identity */
 						pltsql_update_last_identity(seqid, last_identity);
-					}
-				}
-			}
 
-			physical_schema_name = get_namespace_name(tsql_identity_insert.schema_oid);
-
-			seq_options = sequence_options(seqid);
-
-			foreach (seq_lc, seq_options)
-			{
-				DefElem *defel = (DefElem *) lfirst(seq_lc);
+						/*
+						 * We also need to reset the seed.  If the increment
+						 * is positive, we need to find the max identity that
+						 * we've inserted.  Other wise, we need to set the
+						 * min identity.
+						 */
+						seq_options = sequence_options(seqid);
 
-				if (strcmp(defel->defname, "increment") == 0)
-					seq_incr = defGetInt64(defel);
-			}
+						foreach (seq_lc, seq_options)
+						{
+							DefElem *defel = (DefElem *) lfirst(seq_lc);
 
-			/*
-			* Set sequence start value. Since names are qualified (fetched from built-in functions),
-			* double-quote to preserve cases. Second argument of pg_get_serial_sequence() preserves
-			* case by default
-			*/
-			if (seq_incr < 0)
-				setval_query = "SELECT setval(pg_get_serial_sequence($1, $2), sys.get_min_id_from_table($2, $3, $4));";
-			else
-				setval_query = "SELECT setval(pg_get_serial_sequence($1, $2), sys.get_max_id_from_table($2, $3, $4));";
+							if (strcmp(defel->defname, "increment") == 0)
+								do_setval_tsql(seqid, max_identity, true);
+							else
+								do_setval_tsql(seqid, min_identity, true);
```

Comment:
What would be other `defname` values? Is it correct to handle all other cases with `min_identity`?

Reply:
This is actually wrong. I should check the increment.